MonkeyWrench.Web.WebService/Scheduler/Scheduler.cs

```diff
@@ -400,11 +395,43 @@ private static void AddWork (DB db, List<DBHost> hosts, List<DBLane> lanes, List
 								}
 							}
 
-							sql.AppendFormat ("UPDATE RevisionWork SET state = {0} WHERE id = {1} AND state = 10;", (int) (has_dependencies ? DBState.DependencyNotFulfilled : DBState.NotDone), revisionwork.id);
+							sql.AppendFormat ("UPDATE RevisionWork SET state = 11 WHERE id = {0} AND state = 10;", revisionwork.id);
+							if (!editedHostLanes.Contains (hostlane)) editedHostLanes.Add (hostlane);
+						}
+					}
+				}
 
+				foreach (var editedHostLane in editedHostLanes) {
+					int mostRecent = -1;
+					int mostRecentState = -1;
+					using (IDbCommand cmd = db.CreateCommand ()) {
+						cmd.CommandText = @"
+SELECT RevisionWork.id, RevisionWork.state
+FROM RevisionWork
+INNER JOIN Revision ON Revision.id = RevisionWork.revision_id
+WHERE
+     RevisionWork.host_id = @host_id
+ AND RevisionWork.lane_id = @lane_id
+ORDER BY Revision.Date DESC
+LIMIT 1
+";
+						DB.CreateParameter (cmd, "host_id", editedHostLane.host_id);

```

Comment:
Each query is a network trip, since the database is on a different machine, which will add significantly to the total time.

The machines are close by, but right now there are ~340 enabled host lanes, which means worst case scenario is 340 roundtrips to the database machine.

I tried manually filtering on lane_id:

```sql
EXPLAIN ANALYZE SELECT RevisionWork.id, MAX (Revision.date)
FROM RevisionWork
INNER JOIN Revision ON Revision.id = RevisionWork.revision_id
WHERE
RevisionWork.lane_id =3 OR RevisionWork.lane_id =4 OR [...~338 more OR clauses...]
GROUP BY RevisionWork.id;
    Execution time: 3184.721 ms
```

and as you can see it took a bit over 3 seconds. Constructing this query to only include the lane_ids you're actually interested in would make it run proportionately faster, and should not make it slower than your version.
```